pkg/oauthclientscontroller/oauthclientscontroller.go

```diff
@@ -0,0 +1,147 @@
+	authConfigClient configv1client.AuthenticationsGetter
+
+	oauthClientGetter oauthclient.OAuthClientsGetter
+	routeClient       routeclient.RouteInterface
+}
+
+type clientEnsureFunc func(*oauthClientsController, *oauthv1.OAuthClient, *routev1.Route) error
+
+func NewOAuthClientsController(
+	authOpConfigClient authopclient.OperatorClient,
+	authConfigClient configv1client.AuthenticationsGetter,
+	authConfigInformer configinformer.AuthenticationInformer,
+	oauthClientGetter oauthclient.OAuthClientsGetter,
+	oauthClientInformer oauthinformer.OAuthClientInformer,
+	routeClient routeclient.RouteInterface,
+	routeInformer routeinformer.RouteInformer,
+) controller.Runner {
+	c := &oauthClientsController{
+		authOperatorClient: authOpConfigClient,
+
+		authConfigClient: authConfigClient,
+
+		oauthClientGetter: oauthClientGetter,
+		routeClient:       routeClient,
+	}
+
+	targetNameFilter := controller.FilterByNames(nil, authoputil.TargetName)
+	configNameFilter := controller.FilterByNames(nil, authoputil.GlobalConfigName)
+
+	return controller.New("OAuthClientsController", c,
+		controller.WithInformer(oauthClientInformer, controller.FilterFuncs{
+			AddFunc: isBootstrappedClient,
+			UpdateFunc: func(oldObj, newObj metav1.Object) bool {
+				// name change is not possible so we don't have to worry about the clients
+				// disappearing by update -> check only the new object
+				return isBootstrappedClient(newObj)
+			},
+			DeleteFunc: isBootstrappedClient,
+		}),
+		controller.WithInformer(routeInformer, targetNameFilter),
+		controller.WithInformer(authOpConfigClient.Informers.Operator().V1().Authentications(), configNameFilter),
+		controller.WithInformer(authConfigInformer, configNameFilter),
+	)
+}
+
+func (c *oauthClientsController) Key(namespace, name string) (metav1.Object, error) {
+	obj, err := c.oauthClientGetter.OAuthClients().Get(name, metav1.GetOptions{})
+	// the client either got deleted or this is a different object
+	if err != nil && apierrors.IsNotFound(err) {
+		return nil, nil
+	}
+	return obj, err
+}
+
+func (c *oauthClientsController) Sync(obj metav1.Object) error {
+	authOperatorConfig, err := c.authOperatorClient.Client.Authentications().Get(authoputil.GlobalConfigName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	authConfig, err := c.authConfigClient.Authentications().Get(authoputil.GlobalConfigName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Don't do anything if the operator is not in Managed state or the integrated
+	// OAuth server is not supposed to run in the cluster
+	// Eventually we may want to remove the oauthclients in certain states
+	if authOperatorConfig.Spec.ManagementState != operatorv1.Managed ||
+		!isAuthTypeIntegratedOAuthServer(authConfig.Spec.Type) {
+		return nil
+	}
+
+	route, err := c.routeClient.Get(authoputil.TargetName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if obj == nil { // the client was probably deleted -> ensure both clients exist
```

Comment:
This is the wrong construct.  You only reconcile what matches your key `namespace, name`